api/server/sdk/watcher.go

```diff
@@ -0,0 +1,242 @@
+	defer s.Unlock()
+	if s.watchConnections == nil {
+		s.watchConnections = make(map[string][]*watchConnection)
+	}
+	s.watchConnections[eventType] = append(s.watchConnections[eventType], client)
+	logrus.Debugf("successfully register watcher %v", client.name)
+}
+
+func (s *WatcherServer) removeWatcher(name string, eventType string) {
+	s.Lock()
+	defer s.Unlock()
+	var newWatchers []*watchConnection
+	for _, client := range s.watchConnections[eventType] {
+		if client.name == name {
+			// clean up client go channel
+			close(client.eventChannel)
+			continue
+		}
+		newWatchers = append(newWatchers, client)
+	}
+	s.watchConnections[eventType] = newWatchers
+}
+
+func (s *WatcherServer) startWatcher(ctx context.Context, done chan bool) error {
+	group, _ := errgroup.WithContext(ctx)
+	errChan := make(chan error)
+	group.Go(func() error {
+		return s.startVolumeWatcher(ctx, done)
+	})
+
+	// wait for err-group processes to be done
+	go func() {
+		errChan <- group.Wait()
+	}()
+
+	// wait only as long as context deadline allows
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return status.Errorf(codes.Internal, "error starting watcher: %v", err)
+		} else {
+			return nil
+		}
+	case <-ctx.Done():
+		return status.Error(codes.DeadlineExceeded,
+			"Deadline is reached, server side func exiting")
+	}
+}
+
+func (s *WatcherServer) startVolumeWatcher(ctx context.Context, done chan bool) error {
+	if s.watchConnections == nil {
+		s.watchConnections = make(map[string][]*watchConnection)
+	}
+
+	// wait for driver to be initialized with an non-empty volume watcher
+	for {
+		if s.volumeServer.driver(ctx) == nil {
+			continue
+		}
+
+		volumeChannel, _ := s.volumeServer.driver(ctx).GetVolumeWatcher(&api.VolumeLocator{}, make(map[string]string))
+		if volumeChannel == nil {
+			continue
+		}
+		s.volumeChannel = volumeChannel
+		goto volumeWatch
+	}
+	// volumeChannel should be waiting for incoming events
+volumeWatch:
+	for {
+		select {
+		case vol := <-s.volumeChannel:
+			s.RLock()
+			for _, client := range s.watchConnections[volumeEventType] {
+				go client.callBack(vol)
+			}
+			s.RUnlock()
+		case <-done:
+			logrus.Infof("exiting volume watcher\n")
+			break volumeWatch
+		}
+	}
+
+	return nil
+}
+
+func (w *WatcherServer) volumeWatch(
+	req *api.SdkVolumeWatchRequest,
+	stream api.OpenStorageWatch_WatchServer,
+) error {
+	ctx := stream.Context()
+	if w.volumeServer.cluster() == nil || w.volumeServer.driver(ctx) == nil {
+		return status.Error(codes.Unavailable, "Resource has not been initialized")
+	}
+
+	// if input has deadline, ensure graceful exit within that deadline.
+	deadline, ok := ctx.Deadline()
+	var cancel context.CancelFunc
+	if ok {
+		// create a new context that will get done on deadline
+		ctx, cancel = context.WithTimeout(ctx, time.Until(deadline))
+		defer cancel()
+	}
+
+	client := watchConnection{
+		name:         uuid.New(),
+		eventType:    volumeEventType,
+		eventChannel: make(chan interface{}, 2),
```

Comment:
Why create a buffer of only 2? This could lead to issues where for some reason the channel is full (say the stream is blocked) and we keep spinning up go routines on line 132 as part of callback invocation. 

It would be better to have an unbounded channel, so that we keep queueing in the events, and avoid goroutines waiting on sending on this channel.

Reply:
I was referring to an unbounded channel and not unbuffered. Essentially just an unbounded queue that will keep collecting the events.